iothub_client/samples/pnp/common/pnp_device_client_helpers.h

```diff
@@ -6,16 +6,61 @@
 
 #include "iothub_device_client.h"
 
+//
+// Whether we're using a connection string or DPS provisioning for device credentials
+//
+typedef enum PNP_CONNECTION_SECURITY_TYPE_TAG
+{
+    PNP_CONNECTION_SECURITY_TYPE_CONNECTION_STRING,
+    PNP_CONNECTION_SECURITY_TYPE_DPS
+} PNP_CONNECTION_SECURITY_TYPE;
+
+#ifdef USE_PROV_MODULE
+//
+// PNP_DPS_CONFIGURATION is used to configure the DPS device client
+//
+typedef struct PNP_DPS_CONFIGURATION_TAG
+{
+    char* idScope;
+    char* registrationId;
+    char* deviceKey;
+} PNP_DPS_CONFIGURATION;
+#endif /* USE_PROV_MODULE */
+
+//
+// PNP_HELPER_DEVICE_CONFIGURATION is used to setup the IOTHUB_DEVICE_CLIENT_HANDLE
+//
+typedef struct PNP_DEVICE_CONFIGURATION_TAG
+{
+    // Whether we're using connection string or DPS
+    PNP_CONNECTION_SECURITY_TYPE securityType;
+    // The connection string or DPS security information
+    union {
+        char* connectionString;
+#ifdef USE_PROV_MODULE
+        PNP_DPS_CONFIGURATION dpsConfiguration;
+#endif
+    } u;
+    // ModelId of this PnP device
+    const char* modelId;
+    // Whether more verbose tracing is enabled for the IoT Hub client
+    bool enableTracing;
+    // Callback for IoT Hub device methods, which is the mechanism PnP commands use.  If PnP commands
+    // are not used, this should be NULL to conserve memory and bandwidth.
+    IOTHUB_CLIENT_DEVICE_METHOD_CALLBACK_ASYNC deviceMethodCallback;
+    // Callback for IoT Hub device twin notifications, which is the mechanism PnP properties from service use.
+    // If PnP properties are not configured by the server, this should be NULL to conserve memory and bandwidth.
+    IOTHUB_CLIENT_DEVICE_TWIN_CALLBACK deviceTwinCallback;
+} PNP_DEVICE_CONFIGURATION;
+
 //
 // PnPHelper_CreateDeviceClientHandle creates an IOTHUB_DEVICE_CLIENT_HANDLE that will be ready to interact with PnP.
 // Beyond basic handle creation, it also sets the handle to the appropriate ModelId, optionally sets up callback functions
 // for Device Method and Device Twin callbacks (to process PnP Commands and Properties, respectively)
 // as well as some other basic maintenence on the handle. 
 //
-// If the application's model does not implement PnP commands, the application should set deviceMethodCallback to NULL.
-// Anologously, if there are no desired PnP properties, deviceTwinCallback should be NULL.  Not subscribing for these
-// callbacks if they are not needed by the model will save bandwidth and RAM.
+// NOTE: When using DPS based authentication, this function can *block* until DPS responds to the request or timeout.
```

Comment:
does it block until dps responds to a provisioning request (quickish) or the ultimate provisioning result (not quickish)?

Reply:
It's going to block until we get the response from DPS, including iotHub & deviceId with a 60 timeout.